### PR TITLE
Fix issue for nested next page link with singleton containment navigation property

### PR DESCRIPTION
**Pull request by lisicase**
### Issues

This pull request fixes the following issue: OData/AspNetCoreOData#701.

**Issue Summary:** When expanding a singleton (with AutoExpand and Contained attributes) where a contained collection has a limited page size, results are truncated without a nextLink and end promptly after that truncation--without finishing in valid JSON.

### Description

**Fix:** Check for singletons when generating a path to a contained navigation property (so that the singleton is not skipped when looking for the shortest link to said nav property).

**Tests:** Two unit tests & one E2E theory.

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### CODE REVIEW

File: src/Microsoft.AspNet.OData/Microsoft.AspNet.OData.csproj [hunk 1]
```diff
@@ -68,6 +73,12 @@
     <Compile Include="..\CommonAssemblyInfo.cs">
       <Link>Properties\CommonAssemblyInfo.cs</Link>
     </Compile>
+    <Compile Include="..\Microsoft.AspNet.OData.Shared\Common\SRResources1.Designer.cs">
```
Comment (xuzhg) [inline src/Microsoft.AspNet.OData/Microsoft.AspNet.OData.csproj:76]: remove this also.

File: src/Microsoft.AspNet.OData/Microsoft.AspNet.OData.csproj [hunk 2]
```diff
@@ -42,6 +42,11 @@
       <HintPath>..\..\sln\packages\Microsoft.Spatial.7.12.2\lib\net45\Microsoft.Spatial.dll</HintPath>
       <Private>True</Private>
     </Reference>
+    <Reference Include="Newtonsoft.Json, Version=13.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed, processorArchitecture=MSIL" />
```
Comment (xuzhg) [inline src/Microsoft.AspNet.OData/Microsoft.AspNet.OData.csproj:45]: rebase to the latest master branch

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Singleton/SingletonTest.cs [hunk 3]
```diff
@@ -374,7 +427,7 @@ public async Task SingletonDerivedTypeTest(string model, string singletonName, s
         [InlineData("conAttr", "MonstersInc")]
         public async Task SingletonQueryOptionsTest(string model, string singletonName)
         {
-            string requestUri = string.Format(this.BaseAddress + "/{0}/{1}", model, singletonName);
+            string requestUri = string.Format(this.BaseAddress + "/{0}/{1}", model, singletonName); // "http://LAPTOP-4K4FHFF3:11003/conAttr/MonstersInc"
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Singleton/SingletonTest.cs:430]: different people have different machine name.
So, the comment here doesn't make sense. Remove it.

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/LinkGenerationHelpersTest.cs [hunk 4]
```diff
@@ -93,6 +94,60 @@ public void GenerateNavigationLink_WorksToGenerateExpectedNavigationLink_ForSing
...
+        {
+            // Arrange
+            ResourceContext orderItemResource = GetOrderItemResourceForNewSingletonContainer();
+
+            // Act
+            IList<ODataPathSegment> newPaths = orderItemResource.GenerateBaseODataPathSegments();
+
+            // Assert
+            Assert.Equal(3, newPaths.Count);
+            Assert.IsType<Microsoft.OData.UriParser.SingletonSegment>(newPaths[0]); // VipOrder
```
Comment (xuzhg) [inline test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/LinkGenerationHelpersTest.cs:132]: use the "using Microsoft.OData.UriParser.SingletonSegment" at the top and remove the namespace here.

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/LinkGenerationHelpersTest.cs [hunk 5]
```diff
@@ -93,6 +94,60 @@ public void GenerateNavigationLink_WorksToGenerateExpectedNavigationLink_ForSing
...
+            Assert.IsType<Microsoft.OData.UriParser.NavigationPropertySegment>(newPaths[1]); // OrderItems
+            Assert.IsType<Microsoft.OData.UriParser.KeySegment>(newPaths[2]); // 21
```
Comment (xuzhg) [inline test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/LinkGenerationHelpersTest.cs:134]: 21 ?

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/TestModels/OrderItem.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
+﻿//-----------------------------------------------------------------------------
+// <copyright file="OrderLine.cs" company=".NET Foundation">
```
Comment (xuzhg) [inline test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/TestModels/OrderItem.cs:2]: update the copyright to refresh the content

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/TestModels/OrderItemDetail.cs [hunk 7]
```diff
@@ -0,0 +1,18 @@
+﻿//-----------------------------------------------------------------------------
+// <copyright file="OrderLine.cs" company=".NET Foundation">
```
Comment (xuzhg) [inline test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/TestModels/OrderItemDetail.cs:2]: same here

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/TestModels/OrderLine.cs [hunk 8]
```diff
@@ -5,6 +5,9 @@
 // </copyright>
 //------------------------------------------------------------------------------
 
+using Microsoft.AspNet.OData.Builder;
+using System.Collections.Generic;
```
Comment (xuzhg) [inline test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/TestModels/OrderLine.cs:9]: put System* ahead and then orderby

File: src/Microsoft.AspNet.OData/Microsoft.AspNet.OData.csproj (hunk 1 above)
Comment (lisicase) [reply src/Microsoft.AspNet.OData/Microsoft.AspNet.OData.csproj]: ✅

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Singleton/SingletonTest.cs (hunk 3 above)
Comment (lisicase) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Singleton/SingletonTest.cs:430]: ✅

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/LinkGenerationHelpersTest.cs (hunk 5 above)
Comment (lisicase) [reply test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/LinkGenerationHelpersTest.cs:134]: 21 is the expected ID for this path segment

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/LinkGenerationHelpersTest.cs (hunk 4 above)
Comment (lisicase) [reply test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/LinkGenerationHelpersTest.cs:132]: ✅

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/TestModels/OrderLine.cs (hunk 8 above)
Comment (lisicase) [reply test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/TestModels/OrderLine.cs:9]: ✅

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/TestModels/OrderItem.cs (hunk 6 above)
Comment (lisicase) [reply test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/TestModels/OrderItem.cs:2]: ✅

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/TestModels/OrderItemDetail.cs (hunk 7 above)
Comment (lisicase) [reply test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/TestModels/OrderItemDetail.cs:2]: ✅

File: src/Microsoft.AspNet.OData/Microsoft.AspNet.OData.csproj (hunk 2 above)
Comment (lisicase) [reply src/Microsoft.AspNet.OData/Microsoft.AspNet.OData.csproj:45]: ✅

**Comment by lisicase** [comment]
@microsoft-github-policy-service agree [company="microsoft"]

**Comment by lisicase** [comment]
@microsoft-github-policy-service agree company="microsoft"

File: src/Microsoft.AspNet.OData/Microsoft.AspNet.OData.csproj [hunk 9]
```diff
@@ -42,6 +42,11 @@
...
+    <Reference Include="Microsoft.Spatial, Version=7.11.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (xuzhg) [inline src/Microsoft.AspNet.OData/Microsoft.AspNet.OData.csproj:46]: This line is still wrong. You can see Line 42, we have 7.12.2 referencing, but Line46 have 7.11.1 referencing again. 
I think you should rebase your PR to the latest master codes.

Same for newtonsoft.Json package reference.

**Review by xuzhg — CHANGES_REQUESTED** [review]

**Review by xuzhg — APPROVED** [review]

File: src/Microsoft.AspNet.OData/Microsoft.AspNet.OData.csproj (hunk 9 above)
Comment (lisicase) [reply src/Microsoft.AspNet.OData/Microsoft.AspNet.OData.csproj]: ✅

**Pull request merged**